### PR TITLE
Implementation of allow free report transaction

### LINKED ISSUE #2370: Research how to modify node that can receive report tx with free gas price
Conduct with @sorawit to find how to allow report transaction that has a gas price below config.

- [ ] Accept every report transaction
- [ ] Accept only valid transaction (come from valid validator)

### LINKED ISSUE #2372: Implementation of allow free report transaction

### PR DESCRIPTION
**Pull request by Benzbeeb**
Fixed: #2372 
Fixed: #2370 

## Implementation details
Implement AnteHandle for NewMempoolFeeDecorator to allow report gas fee

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [x] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [x] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [x] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: chain/x/ante/ante.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package ante
```
Comment (taobun) [inline chain/x/ante/ante.go]: Move to /x/oracle/ante/ante.go

File: chain/x/ante/ante.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+import (
+	"github.com/bandprotocol/bandchain/chain/x/oracle"
```
Comment (taobun) [inline chain/x/ante/ante.go]: Order import like other files.

File: chain/x/ante/ante.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+		}
+	}
+	return true
+}
+
+func BandWrapAnteHandler(ante sdk.AnteHandler, oracleKeeper oracle.Keeper) sdk.AnteHandler {
+	return func(ctx sdk.Context, tx sdk.Tx, simulate bool) (newCtx sdk.Context, err error) {
+		isValidReportTx := true
+		for _, msg := range tx.GetMsgs() {
+			if isValidReportTx = checkValidReportMsg(ctx, oracleKeeper, msg); !isValidReportTx {
```
Comment (taobun) [inline chain/x/ante/ante.go]: I thought set `isValidReportTx` to false in if condition looks better.

File: chain/app/app.go [hunk 4]
```diff
@@ -237,7 +238,8 @@ func NewBandApp(
 	// initialize BaseApp.
 	app.SetInitChainer(app.InitChainer)
 	app.SetBeginBlocker(app.BeginBlocker)
-	app.SetAnteHandler(ante.NewAnteHandler(app.AccountKeeper, app.SupplyKeeper, ante.DefaultSigVerificationGasConsumer))
+	wrappedAnte := bandante.BandWrapAnteHandler(ante.NewAnteHandler(app.AccountKeeper, app.SupplyKeeper, auth.DefaultSigVerificationGasConsumer), app.OracleKeeper)
```
Comment (taobun) [inline chain/app/app.go]: Create a flag for toggle wrap or unwrap mode

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/emitter/app.go [hunk 5]
```diff
@@ -41,11 +41,11 @@ type App struct {
 func NewBandAppWithEmitter(
 	kafkaURI string, logger log.Logger, db dbm.DB, traceStore io.Writer, loadLatest bool,
 	invCheckPeriod uint, skipUpgradeHeights map[int64]bool, home string,
-	baseAppOptions ...func(*bam.BaseApp),
+	withFeeReportTx bool, baseAppOptions ...func(*bam.BaseApp),
```
Comment (sorawit) [inline chain/emitter/app.go]: Doesn't seem correct

File: chain/x/oracle/ante/ante.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+		if !keeper.ContainsEID(req.RawRequests, report.ExternalID) {
+			return false
+		}
+	}
+	return true
+}
+
+// BandWrapAnteHandler returns a new ante handler that set minimum gas price to be zero if the
+// incoming tx is a valid report transaction.
+func BandWrapAnteHandler(ante sdk.AnteHandler, oracleKeeper oracle.Keeper) sdk.AnteHandler {
```
Comment (sorawit) [inline chain/x/oracle/ante/ante.go]: `NewFeelessReportsAnteHandler`

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/oracle/ante/ante.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+	}
+	for _, report := range rep.RawReports {
+		if !keeper.ContainsEID(req.RawRequests, report.ExternalID) {
+			return false
+		}
+	}
+	return true
+}
+
+// NewFeelessReportsAnteHandler returns a new ante handler that set minimum gas price to be zero if the
```
Comment (sorawit) [inline chain/x/oracle/ante/ante.go]: sets ... that waive minimum gas price requirement if

**Review by sorawit — APPROVED** [review]
lgtm. just some minor comment change

**Pull request merged**